mp/game/momentum/resource/momentum_english_ref.res

```diff
@@ -627,7 +627,20 @@
         "MOM_MB_Safeguard_Map_Quit_ToMenu_Msg" "You are trying to quit to menu while your timer is running!\nAre you sure you want to quit?\n(You can go to Settings -> Gameplay -> Safeguards to turn this off!)"
 
         "MOM_Momentum" "Momentum"
-        
+
+        // Safeguard warnings
+        "MOM_Safeguard_Movement_Keys_Warning_Practice"  "You cannot enable practice mode when movement keys are held down while the timer is running!\nYou can turn this safeguard off in the Gameplay Settings!"
+        "MOM_Safeguard_Movement_Keys_Warning_Restart"   "You cannot restart map when movement keys are held down while the timer is running!\nYou can turn this safeguard off in the Gameplay Settings!"
+        "MOM_Safeguard_Movement_Keys_Warning_TeleSave"  "You cannot teleport to saveloc when movement keys are held down while the timer is running!\nYou can turn this safeguard off in the Gameplay Settings!"
+        "MOM_Safeguard_Movement_Keys_Warning_Chat"      "You cannot open chat when movement keys are held down while the timer is running!\nYou can turn this safeguard off in the Gameplay Settings!"
+        "MOM_Safeguard_Movement_Keys_Warning_TeleStage" "You cannot teleport to a stage when movement keys are held down while the timer is running!\nYou can turn this safeguard off in the Gameplay Settings!"
+
+        "MOM_Safeguard_Double_Press_Warning_Practice"  "You must double press the key to enable practice mode while the timer is running!\nYou can turn this safeguard off in the Gameplay Settings!"
+        "MOM_Safeguard_Double_Press_Warning_Restart"   "You must double press the key to restart map while the timer is running!\nYou can turn this safeguard off in the Gameplay Settings!"
+        "MOM_Safeguard_Double_Press_Warning_TeleSave"  "You must double press the key to teleport to saveloc while the timer is running!\nYou can turn this safeguard off in the Gameplay Settings!"
+        "MOM_Safeguard_Double_Press_Warning_Chat"      "You must double press the key to open chat while the timer is running!\nYou can turn this safeguard off in the Gameplay Settings!"
+        "MOM_Safeguard_Double_Press_Warning_TeleStage" "You must double press the key to teleport to a stage while the timer is running!\nYou can turn this safeguard off in the Gameplay Settings!"
```

Comment:
```suggestion
         "MOM_Safeguard_Type_Practice_Mode" "enable practice mode"
         "MOM_Safeguard_Type_Restart" "restart map"
         . . . etc do the rest of the types here
         
         "MOM_Safeguard_Movement_Keys_Warning" "You cannot %s1 when movement keys are held down . . ."
         "MOM_Safeguard_Double_Press_Warning" "You must double press the bound key to %s1 while the timer . . ."
```

Then each safeguard stores the token name of the type that it is in the constructor.

Then comes the fun part, displaying.

In a traditional sense, to get at the `%s1` in the string, you'd do like `CConstructLocalizedString(g_pVguiLocalize->FindSafe("#MOM_Safeguard_Movement_Keys_Warning"), safeguard->GetTypeString())` and have the `GetTypeString()` just return the `wchar_t*` that's assigned to each safeguard instance (e.g. `new CRunSafeguard(L"MOM_Safeguard_Type_Restart")`).

However, should this be only in the server (GAME_DLL) code? Opening the chat, for example, is done entirely through the client so I don't think this even works for that. Something you could do is create a `ShowSafeguardHUDHint` function that's shared between server and client, and on the server it sends a user message, and on the client it registers and listens for it and does the actual hud hint showing. It could then just be called directly on the client.

I think this also calls for a globalized hint panel that's more of like a message stack panel, that, sorta like the chat panel does, takes messages and appends them to the bottom of a panel, with each new message being added above it. Each message would fade out, and as ones do, they get removed and every message animated down one slot. This would be an entirely different PR though.

Reply:
> on the server it sends a user message  
> client registers and listens for it  

With the usermessages, doesn't the client need a whole separate function to be called when listening? Thus meaning that sharing the `ShowSafeguardHUDHint` function would be pointless if the server is sending out that usermessage because listening causes it to call a new function?  

I'm just really confused honestly, this is what I have so far:
![image](https://user-images.githubusercontent.com/33820042/97952546-ca705e00-1d6b-11eb-9ea3-37adf70a8e87.png)

Basically, my understanding here is that upon `ShowSafeguardHUDHint` being called, the client side would draw the hint
The server would then make its usermessage, causing the client side to directly recall `ShowSafeguardHUDHint`, making it re-draw the hint unnecessarily

But if I were to handle drawing the hint inside of the `MsgFunc` listener function, it would only be called once properly (versus client side `ShowSafeguardHUDHint`'s twice), but would then mean that `ShowSafeguardHUDHint` would effectively be a server only function, going against what you said in the message

Reply:
The thing about this is, you're creating these safeguards on both the server and the client. What you're doing with this server message and client display architecture is:

Server calls "display" but actually can't do it because it doesn't have VGUI on it, so it needs to network to the client to do it. The UTIL_ShowHudHintText did this for you but I think we're misusing it, hence my suggestion. What we're doing instead is creating our own usermessage that the SafeguardSystem should hook into to display the correct safeguard hint. You could even pass the safeguard enum type to this message so it can easily do it.

The client then calls Display, and because of the ifdef code, only does the actual displaying, and not the usermessage stuff. The usermessage callback shouldn't be in the safeguard itself, but rather the system. Safeguards should only have a Display function.

If it's called on the client, only the client code is run. If it's called on the server, only the server code is called.